Error: JPEG error: marker was not found. Scanned from Samsung MFP.

Link to PDF file (or attach file here): [scan.pdf](https://github.com/mozilla/pdf.js/files/865245/scan.pdf)

Configuration:
- Web browser and its version: Chrome 56.0.2924.87 (64-bit) and Firefox 52.0.1 (64-bit)
- Operating system and its version: Windows 10 64 bit and Ubuntu 16.10
- PDF.js version: Stable 1.6.210 and Beta 1.7.225
- Is an extension: No

Steps to reproduce the problem:
1. Upload the pdf to the Online demo site https://mozilla.github.io/pdf.js/web/viewer.html
2. All pages are blank

What is the expected behavior? PDF will be rendered properly 
![expected](https://cloud.githubusercontent.com/assets/4175765/24256036/1babfc80-1022-11e7-9ca1-ef7d1a68aa10.PNG)


What went wrong? There are pages but no content.  In the developer console of the browser this is the logs:
`PDF a650df87b150696f9e4415aeeaece60c [1.4 CLX-8380 Series / 300] (PDF.js: 1.7.378)
pdf.worker.js:345 Error: JPEG error: marker was not found
/pdf.js/build/pdf.worker.js:345 Error: JPEG error: marker was not found
pdf.worker.js:346     at error (https://mozilla.github.io/pdf.js/build/pdf.worker.js:346:15)
    at decodeScan (https://mozilla.github.io/pdf.js/build/pdf.worker.js:44260:5)
    at JpegImage.parse (https://mozilla.github.io/pdf.js/build/pdf.worker.js:44637:22)
    at JpegStream_ensureBuffer [as ensureBuffer] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:3087:13)
    at JpegStream_getBytes [as getBytes] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:3094:8)
    at PDFImage_getImageBytes [as getImageBytes] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:42687:22)
    at PDFImage_createImageData [as createImageData] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:42619:26)
    at https://mozilla.github.io/pdf.js/build/pdf.worker.js:25799:28
/pdf.js/build/pdf.worker.js:346     at error (https://mozilla.github.io/pdf.js/build/pdf.worker.js:346:15)
    at decodeScan (https://mozilla.github.io/pdf.js/build/pdf.worker.js:44260:5)
    at JpegImage.parse (https://mozilla.github.io/pdf.js/build/pdf.worker.js:44637:22)
    at JpegStream_ensureBuffer [as ensureBuffer] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:3087:13)
    at JpegStream_getBytes [as getBytes] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:3094:8)
    at PDFImage_getImageBytes [as getImageBytes] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:42687:22)
    at PDFImage_createImageData [as createImageData] (https://mozilla.github.io/pdf.js/build/pdf.worker.js:42619:26)
    at https://mozilla.github.io/pdf.js/build/pdf.worker.js:25799:28
`
![actual](https://cloud.githubusercontent.com/assets/4175765/24256164/59f7429c-1022-11e7-862b-5236468aa597.PNG)

I think this is also related with #8090 and #8104

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension): Tested using mozilla.github.io/pdf.js

--- a/src/core/jpg.js
+++ b/src/core/jpg.js
@@ -323,7 +323,7 @@ var JpegImage = (function JpegImageClosure() {
       decodeFn = decodeBaseline;
     }
 
-    var mcu = 0, marker;
+    var mcu = 0, fileMarker;
     var mcuExpected;
     if (componentsLength === 1) {
       mcuExpected = components[0].blocksPerLine * components[0].blocksPerColumn;
@@ -365,14 +365,16 @@ var JpegImage = (function JpegImageClosure() {
 
       // find marker
       bitsCount = 0;
-      marker = (data[offset] << 8) | data[offset + 1];
-      // Some bad images seem to pad Scan blocks with zero bytes, skip past
+      fileMarker = findNextFileMarker(data, offset);
+      // Some bad images seem to pad Scan blocks with e.g. zero bytes, skip past
       // those to attempt to find a valid marker (fixes issue4090.pdf).
-      while (data[offset] === 0x00 && offset < data.length - 1) {
-        offset++;
-        marker = (data[offset] << 8) | data[offset + 1];
+      if (fileMarker && fileMarker.invalid) {
+        warn('decodeScan - unexpected MCU data, next marker is: ' +
+             fileMarker.invalid);
+        offset = fileMarker.offset;
       }
-      if (marker <= 0xFF00) {
+      var marker = fileMarker && fileMarker.marker;
+      if (!marker || marker <= 0xFF00) {
         error('JPEG error: marker was not found');
       }
 
@@ -381,6 +383,15 @@ var JpegImage = (function JpegImageClosure() {
       } else {
         break;
       }
+    }
+
+    fileMarker = findNextFileMarker(data, offset);
+    // Some images include more Scan blocks than expected, skip past those and
+    // attempt to find the next valid marker (fixes issue8182.pdf).
+    if (fileMarker && fileMarker.invalid) {
+      warn('decodeScan - unexpected Scan data, next marker is: ' +
+           fileMarker.invalid);
+      offset = fileMarker.offset;
     }
 
     return offset - startOffset;
@@ -594,6 +605,39 @@ var JpegImage = (function JpegImageClosure() {
     return a <= 0 ? 0 : a >= 255 ? 255 : a;
   }
 
+  function findNextFileMarker(data, currentPos, startPos) {
+    function peekUint16(pos) {
+      return (data[pos] << 8) | data[pos + 1];
+    }
+
+    var maxPos = data.length - 1;
+    var newPos = startPos < currentPos ? startPos : currentPos;
+
+    if (currentPos >= maxPos) {
+      return null; // Don't attempt to read non-existent data and just return.
+    }
+    var currentMarker = peekUint16(currentPos);
+    if (currentMarker >= 0xFFC0 && currentMarker <= 0xFFFE) {
+      return {
+        invalid: null,
+        marker: currentMarker,
+        offset: currentPos,
+      };
+    }
+    var newMarker = peekUint16(newPos);
+    while (!(newMarker >= 0xFFC0 && newMarker <= 0xFFFE)) {
+      if (++newPos >= maxPos) {
+        return null; // Don't attempt to read non-existent data and just return.
+      }
+      newMarker = peekUint16(newPos);
+    }
+    return {
+      invalid: currentMarker.toString(16),
+      marker: newMarker,
+      offset: newPos,
+    };
+  }
+
   JpegImage.prototype = {
     parse: function parse(data) {
 
@@ -604,25 +648,14 @@ var JpegImage = (function JpegImageClosure() {
       }
 
       function readDataBlock() {
-        function isValidMarkerAt(pos) {
-          if (pos < data.length - 1) {
-            return (data[pos] === 0xFF &&
-                    data[pos + 1] >= 0xC0 && data[pos + 1] <= 0xFE);
-          }
-          return true;
-        }
-
         var length = readUint16();
         var endOffset = offset + length - 2;
 
-        if (!isValidMarkerAt(endOffset)) {
+        var fileMarker = findNextFileMarker(data, endOffset, offset);
+        if (fileMarker && fileMarker.invalid) {
           warn('readDataBlock - incorrect length, next marker is: ' +
-               (data[endOffset] << 8 | data[endOffset + 1]).toString('16'));
-          var pos = offset;
-          while (!isValidMarkerAt(pos)) {
-            pos++;
-          }
-          endOffset = pos;
+               fileMarker.invalid);
+          endOffset = fileMarker.offset;
         }
 
         var array = data.subarray(offset, endOffset);

--- a/test/pdfs/issue8182.pdf.link
+++ b/test/pdfs/issue8182.pdf.link
@@ -1,0 +1,1 @@
+https://github.com/mozilla/pdf.js/files/861340/PDF.with.blank.pages.pdf

--- a/test/test_manifest.json
+++ b/test/test_manifest.json
@@ -740,6 +740,14 @@
        "link": true,
        "type": "eq"
     },
+    {  "id": "issue8182",
+       "file": "pdfs/issue8182.pdf",
+       "md5": "e295ae13dcbefd449f9a4957aed5e582",
+       "rounds": 1,
+       "link": true,
+       "lastPage": 1,
+       "type": "eq"
+    },
     {  "id": "txt2pdf",
        "file": "pdfs/txt2pdf.pdf",
        "md5": "02cefa0f5e8d96313bb05163b2f88c8c",

